google/resource_endpoints_service.go

```diff
@@ -0,0 +1,297 @@
+	// new config.  A ConfigSource contains a ConfigFile which contains the b64
+	// encoded contents of the file.  OpenAPI requires only one file.
+	configfile := servicemanagement.ConfigFile{
+		FileContents: base64.StdEncoding.EncodeToString([]byte(config_text)),
+		FileType:     "OPEN_API_YAML",
+		FilePath:     "heredoc.yaml",
+	}
+	return servicemanagement.ConfigSource{
+		Files: []*servicemanagement.ConfigFile{&configfile},
+	}
+}
+
+func getGrpcConfigSource(service_config, proto_config string) servicemanagement.ConfigSource {
+	// gRPC requires both the file specifying the service and the compiled protobuf,
+	// but they can be in any order.
+	yml_configfile := servicemanagement.ConfigFile{
+		FileContents: base64.StdEncoding.EncodeToString([]byte(service_config)),
+		FileType:     "SERVICE_CONFIG_YAML",
+		FilePath:     "heredoc.yaml",
+	}
+	proto_configfile := servicemanagement.ConfigFile{
+		FileContents: base64.StdEncoding.EncodeToString([]byte(proto_config)),
+		FileType:     "FILE_DESCRIPTOR_SET_PROTO",
+		FilePath:     "api_def.pb",
+	}
+	return servicemanagement.ConfigSource{
+		Files: []*servicemanagement.ConfigFile{&yml_configfile, &proto_configfile},
+	}
+}
+
+func resourceEndpointsServiceCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+	// If the service doesn't exist, we'll need to create it, but if it does, it
+	// will be reused.
+	serviceName := d.Get("service_name").(string)
+	servicesService := servicemanagement.NewServicesService(config.clientServiceMan)
+	_, err = servicesService.Get(serviceName).Do()
+	if err != nil {
+		_, err = servicesService.Create(&servicemanagement.ManagedService{ProducerProjectId: project, ServiceName: serviceName}).Do()
+		if err != nil {
+			return err
+		}
+	}
+	// Do a rollout using the update mechanism.
+	err = resourceEndpointsServiceUpdate(d, meta)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(serviceName)
+	return resourceEndpointsServiceRead(d, meta)
+}
+
+func resourceEndpointsServiceUpdate(d *schema.ResourceData, meta interface{}) error {
+	// This update is not quite standard for a terraform resource.  Instead of
+	// using the go client library to send an HTTP request to update something
+	// serverside, we have to push a new configuration, wait for it to be
+	// parsed and loaded, then create and push a rollout and wait for that
+	// rollout to be completed.
+	// There's a lot of moving parts there, and all of them have knobs that can
+	// be tweaked if the user is using gcloud.  In the interest of simplicity,
+	// we currently only support full rollouts - anyone trying to do incremental
+	// rollouts or A/B testing is going to need a more precise tool than this resource.
+	config := meta.(*Config)
+	serviceName := d.Get("service_name").(string)
+	openapi_config, ok := d.GetOk("openapi_config")
+	var source servicemanagement.ConfigSource
+	if ok {
```

Comment:
you could do `if openapiConfig, ok := d.GetOk("openapi_config"); ok {` here since that var is only used within the scope of the if/else

Reply:
Done.